### ISSUE TITLE
LWG-4164 Missing guarantees for `forward_list` modifiers

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#5126]
LWG-4164 Missing guarantees for `forward_list` modifiers

Even if we already conform here, we should add test coverage for these increased guarantees (particularly strong EH).

**Comment by ovatonne** [comment]
I investigated this a bit

There is some coverage in `libcxx/test/std/containers/sequences/forwardlist/exception_safety.pass.cpp` added for [LWG-4164](https://cplusplus.github.io/LWG/issue4164) in [llvm-project#131025](https://github.com/llvm/llvm-project/pull/131025), but I think the test is skipped on CI. It contains missing TODOs, but covers at least the mandated modifiers from [LWG-4164](https://cplusplus.github.io/LWG/issue4164). Unfortunately, the test is skipped on CI but I didn't fully understand why (is it the `// UNSUPPORTED: c++03, no-exceptions` ?).

When you created this issue, was the goal to add tests only for `forward_list`'s modifiers under this repo or did you expect more ?

I'd gladly take this issue to get used to the testing infrastructure. If this is not on your priority list, I can investigate other issues though

**Comment by StephanTLavavej** [comment]
> Unfortunately, the test is skipped on CI but I didn't fully understand why

The test is running and passing:

```
D:\GitHub\STL\out\x64>python tests\utils\stl-lit\stl-lit.py -o testing_x64.log -Dnotags=ASAN --filter=exception_safety %STL%\llvm-project\libcxx\test\std\containers\sequences\forwardlist
-- Testing: 6 of 252 tests, 6 workers --
SKIPPED: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:1 (1 of 6)
SKIPPED: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:1 (2 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:0 (3 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:2 (4 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:0 (5 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:2 (6 of 6)

Testing Time: 3.43s

Total Discovered Tests: 252
  Excluded: 246 (97.62%)
  Skipped :   2 (0.79%)
  Passed  :   4 (1.59%)
```

Configuration 0 is vanilla MSVC, 1 is ASan MSVC, 2 is Clang. Observe that for `std/containers/sequences/forwardlist/exception_safety.pass.cpp`, `:0` and `:2` are passing, while `:1` is skipped because I passed `-Dnotags=ASAN`. If I omit that, they all run and pass:

```
D:\GitHub\STL\out\x64>python tests\utils\stl-lit\stl-lit.py -o testing_x64.log --filter=exception_safety %STL%\llvm-project\libcxx\test\std\containers\sequences\forwardlist
-- Testing: 6 of 252 tests, 6 workers --
PASS: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:0 (1 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:1 (2 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:2 (3 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/forwardlist.modifiers/push_front_exception_safety.pass.cpp:2 (4 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:1 (5 of 6)
PASS: libc++ :: std/containers/sequences/forwardlist/exception_safety.pass.cpp:0 (6 of 6)

Testing Time: 13.29s

Total Discovered Tests: 252
  Excluded: 246 (97.62%)
  Passed  :   6 (2.38%)
```

> (is it the // UNSUPPORTED: c++03, no-exceptions ?).

No, those aren't a factor here. We don't have a C++03 mode, and we always have exception support (when testing libcxx). Some UNSUPPORTED things activate for us, just not these.

If libcxx already has complete coverage for this LWG issue, I would be comfortable with resolving it as "nothing to do". If we have to add test coverage, we do that in `tests/std`, but it's fine to rely on libcxx when we don't specifically need to exercise Standard modes or various compiler options.

**Comment by ovatonne** [comment]
Ok, I completely misread the logs and didn't understand that. Thank you for the explanation ! I guess I'm still missing some knowledge on the test infrastructure, but at least much more things make sense now.

---

I think it is indeed missing some coverage of this LWG issue unfortunately, as stated in the test's comment
```
// TODO:
// - throwing upon moving;
// - initializer lists;
// - throwing when constructing the element in place.
```

If I'm not mistaken, it is missing some modifiers including some that were added because of the wording change for the whole section:

- `template<class... Args> constexpr reference emplace_front(Args&&... args);`
- `constexpr void push_front(T&& x);`
- `constexpr iterator insert_after(const_iterator position, initializer_list<T> il);`
- `template<class... Args> constexpr iterator emplace_after(const_iterator position, Args&&... args);`

I think the tests are properly written to verify `If an exception is thrown by any of these member functions there is no effect on the container` by using the operator==. I don't see any test in `tests/std` or `libcxx` on iterator/reference invalidation though, which has been reworded by [LWG-4164](https://cplusplus.github.io/LWG/issue4164).

Do you think it is worth adding those missing tests ? I could add the missing modifiers in libcxx if it makes more sense for example.

**Comment by StephanTLavavej** [comment]
It's probably worth adding test coverage, as that will exercise debug checks for iterator invalidation.

If libcxx's test coverage is incomplete, then completing it upstream would be worthwhile, even though we usually add new test coverage just in `tests/std` for ease of development, your call.

(Apologies if I don't notice GitHub notifications; I am severely overloaded and don't always go through them. I'm responsive on the STL Discord along with our other contributors there, though.)

**Comment by ovatonne** [comment]
I agree, I went along the path to add the few missing functions in libcxx first, then probably add iterator non-invalidation tests on this repo (though I didn't think much how to properly test this for now).

(Absolutely no problem, I only had time to setup llvm project on my machine and running libcxx tests from their repo, I don't need faster response ! I will sure join the Discord, I didn't notice it somehow ^^)
